--- attention_processor.py	2024-09-11 19:51:35.404742200 +0800
+++ attention_processor.py	2024-09-11 19:56:20.519589500 +0800
@@ -761,6 +761,9 @@
         hidden_states = torch.bmm(attention_probs, value)
         hidden_states = attn.batch_to_head_dim(hidden_states)
 
+        hidden_states = torch.unsqueeze(hidden_states, 0)
+        hidden_states = torch.squeeze(hidden_states, 0)
+
         # linear proj
         hidden_states = attn.to_out[0](hidden_states, *args)
         # dropout
