diff --git a/egs/librispeech/ASR/zipformer/export-aie-streaming.py b/egs/librispeech/ASR/zipformer/export-aie-streaming.py
index 6bc9b18..b05ce07 100755
--- a/egs/librispeech/ASR/zipformer/export-aie-streaming.py
+++ b/egs/librispeech/ASR/zipformer/export-aie-streaming.py
@@ -65,6 +65,9 @@ See ./onnx_pretrained-streaming.py for how to use the exported ONNX models.

 import argparse
 import logging
+import time
+import json
+
 from pathlib import Path
 from typing import Dict, List, Tuple

@@ -77,6 +80,8 @@ from onnxruntime.quantization import QuantType, quantize_dynamic
 from scaling_converter import convert_scaled_to_non_scaled
 from train import add_model_arguments, get_model, get_params
 from zipformer import Zipformer2
+import mindietorch
+from mindietorch import _enums

 from icefall.checkpoint import (
     average_checkpoints,
@@ -201,8 +206,29 @@ class OnnxEncoder(nn.Module):
     def forward(
         self,
         x: torch.Tensor,
-        states: List[torch.Tensor],
+        state_0, state_1, state_2, state_3, state_4, state_5, state_6, state_7, state_8, state_9, state_10,
+        state_11, state_12, state_13, state_14, state_15, state_16, state_17, state_18, state_19, state_20,
+        state_21, state_22, state_23, state_24, state_25, state_26, state_27, state_28, state_29, state_30,
+        state_31, state_32, state_33, state_34, state_35, state_36, state_37, state_38, state_39, state_40,
+        state_41, state_42, state_43, state_44, state_45, state_46, state_47, state_48, state_49, state_50,
+        state_51, state_52, state_53, state_54, state_55, state_56, state_57, state_58, state_59, state_60,
+        state_61, state_62, state_63, state_64, state_65, state_66, state_67, state_68, state_69, state_70,
+        state_71, state_72, state_73, state_74, state_75, state_76, state_77, state_78, state_79, state_80,
+        state_81, state_82, state_83, state_84, state_85, state_86, state_87, state_88, state_89, state_90,
+        state_91, state_92, state_93, state_94, state_95, state_96, state_97
+        # states: List[torch.Tensor],
     ) -> Tuple[torch.Tensor, torch.Tensor, List[torch.Tensor]]:
+        states = [state_0, state_1, state_2, state_3, state_4, state_5, state_6, state_7, state_8, state_9, state_10,
+        state_11, state_12, state_13, state_14, state_15, state_16, state_17, state_18, state_19, state_20,
+        state_21, state_22, state_23, state_24, state_25, state_26, state_27, state_28, state_29, state_30,
+        state_31, state_32, state_33, state_34, state_35, state_36, state_37, state_38, state_39, state_40,
+        state_41, state_42, state_43, state_44, state_45, state_46, state_47, state_48, state_49, state_50,
+        state_51, state_52, state_53, state_54, state_55, state_56, state_57, state_58, state_59, state_60,
+        state_61, state_62, state_63, state_64, state_65, state_66, state_67, state_68, state_69, state_70,
+        state_71, state_72, state_73, state_74, state_75, state_76, state_77, state_78, state_79, state_80,
+        state_81, state_82, state_83, state_84, state_85, state_86, state_87, state_88, state_89, state_90,
+        state_91, state_92, state_93, state_94, state_95, state_96, state_97]
+
         N = x.size(0)
         T = self.chunk_size * 2 + self.pad_length
         x_lens = torch.tensor([T] * N, device=x.device)
@@ -252,7 +278,8 @@ class OnnxEncoder(nn.Module):
             new_processed_lens,
         ]

-        return encoder_out, new_states
+        new_states.insert(0, encoder_out)
+        return tuple(new_states)

     def get_init_states(
         self,
@@ -273,7 +300,7 @@ class OnnxEncoder(nn.Module):

         states.append(embed_states)

-        processed_lens = torch.zeros(batch_size, dtype=torch.int64, device=device)
+        processed_lens = torch.zeros(batch_size, dtype=torch.int32, device=device)
         states.append(processed_lens)

         return states
@@ -434,6 +461,10 @@ def export_encoder_model_onnx(
     }
     logging.info(f"meta_data: {meta_data}")

+    # save meta data as json file
+    with open('encoder_meta_data.json', 'w') as f:
+        json.dump(meta_data, f)
+
     for i in range(len(init_state[:-2]) // 6):
         build_inputs_outputs(init_state[i * 6 : (i + 1) * 6], i)

@@ -460,23 +491,56 @@ def export_encoder_model_onnx(
     logging.info(input_names)
     logging.info(output_names)

-    torch.onnx.export(
-        encoder_model,
-        (x, init_state),
-        encoder_filename,
-        verbose=False,
-        opset_version=opset_version,
-        input_names=input_names,
-        output_names=output_names,
-        dynamic_axes={
-            "x": {0: "N"},
-            "encoder_out": {0: "N"},
-            **inputs,
-            **outputs,
-        },
-    )
+    encoder_model.eval()
+
+    print('X shape: ', x.shape)
+    inputs = [x]  # torch.Size([1, 45, 80])
+    for state in init_state:
+        inputs.append(state)

-    add_meta_data(filename=encoder_filename, meta_data=meta_data)
+    ts_model = torch.jit.trace(encoder_model, tuple(inputs))
+    print('Done tracing encoder model.')
+    ts_model.save(encoder_filename)
+
+    mindietorch.set_device(0)
+    model = torch.jit.load(encoder_filename)
+    model.eval()
+
+    input_info = []
+    input_info.append(mindietorch.Input(x.shape))
+    for state in init_state[:-1]:
+        input_info.append(mindietorch.Input(state.shape))
+    input_info.append(mindietorch.Input(init_state[-1].shape, dtype=torch.int32))
+
+    print('Start compiling model...')
+    compiled_model = mindietorch.compile(
+        model,
+        inputs=input_info,
+        precision_policy=_enums.PrecisionPolicy.FP32,
+        soc_version="Ascend310P3")
+    print('Model compiled successfully.')
+    save_path = str(encoder_filename).replace('.pt', '_aie.pt')
+    compiled_model.save(save_path)
+
+    """ONNX export"""
+    # encoder_filename = str(encoder_filename).replace('.pt', '.onnx')
+    # torch.onnx.export(
+    #     encoder_model,
+    #     (x, init_state),
+    #     encoder_filename,
+    #     verbose=False,
+    #     opset_version=opset_version,
+    #     input_names=input_names,
+    #     output_names=output_names,
+    #     dynamic_axes={
+    #         "x": {0: "N"},
+    #         "encoder_out": {0: "N"},
+    #         **inputs,
+    #         **outputs,
+    #     },
+    # )
+    #
+    # add_meta_data(filename=encoder_filename, meta_data=meta_data)


 def export_decoder_model_onnx(
@@ -505,27 +569,53 @@ def export_decoder_model_onnx(
     context_size = decoder_model.decoder.context_size
     vocab_size = decoder_model.decoder.vocab_size

-    y = torch.zeros(10, context_size, dtype=torch.int64)
-    decoder_model = torch.jit.script(decoder_model)
-    torch.onnx.export(
-        decoder_model,
-        y,
-        decoder_filename,
-        verbose=False,
-        opset_version=opset_version,
-        input_names=["y"],
-        output_names=["decoder_out"],
-        dynamic_axes={
-            "y": {0: "N"},
-            "decoder_out": {0: "N"},
-        },
-    )
-
     meta_data = {
         "context_size": str(context_size),
         "vocab_size": str(vocab_size),
     }
-    add_meta_data(filename=decoder_filename, meta_data=meta_data)
+
+    # save meta data as json file
+    with open('decoder_meta_data.json', 'w') as f:
+        json.dump(meta_data, f)
+
+    y = torch.zeros(1, context_size, dtype=torch.int32)
+    # trace the model
+    decoder_model.eval()
+    trace_model = torch.jit.trace(decoder_model, y)
+    trace_model.save(decoder_filename)
+
+    mindietorch.set_device(0)
+    model = torch.jit.load(decoder_filename)
+    model.eval()
+
+    input_info = [mindietorch.Input(y.shape, dtype=torch.int32)]
+    print('Start compiling model...')
+    compiled_model = mindietorch.compile(
+        model,
+        inputs=input_info,
+        precision_policy=_enums.PrecisionPolicy.FP16,
+        soc_version="Ascend310P3")
+    print('Decoder compiled successfully.')
+
+    save_path = str(decoder_filename).replace('.pt', '_aie.pt')
+    compiled_model.save(save_path)
+
+    # decoder_model = torch.jit.script(decoder_model)
+    # torch.onnx.export(
+    #     decoder_model,
+    #     y,
+    #     decoder_filename,
+    #     verbose=False,
+    #     opset_version=opset_version,
+    #     input_names=["y"],
+    #     output_names=["decoder_out"],
+    #     dynamic_axes={
+    #         "y": {0: "N"},
+    #         "decoder_out": {0: "N"},
+    #     },
+    # )
+
+    # add_meta_data(filename=decoder_filename, meta_data=meta_data)


 def export_joiner_model_onnx(
@@ -546,30 +636,55 @@ def export_joiner_model_onnx(
     joiner_dim = joiner_model.output_linear.weight.shape[1]
     logging.info(f"joiner dim: {joiner_dim}")

-    projected_encoder_out = torch.rand(11, joiner_dim, dtype=torch.float32)
-    projected_decoder_out = torch.rand(11, joiner_dim, dtype=torch.float32)
+    projected_encoder_out = torch.rand(1, joiner_dim, dtype=torch.float32)
+    projected_decoder_out = torch.rand(1, joiner_dim, dtype=torch.float32)

-    torch.onnx.export(
-        joiner_model,
-        (projected_encoder_out, projected_decoder_out),
-        joiner_filename,
-        verbose=False,
-        opset_version=opset_version,
-        input_names=[
-            "encoder_out",
-            "decoder_out",
-        ],
-        output_names=["logit"],
-        dynamic_axes={
-            "encoder_out": {0: "N"},
-            "decoder_out": {0: "N"},
-            "logit": {0: "N"},
-        },
-    )
-    meta_data = {
-        "joiner_dim": str(joiner_dim),
-    }
-    add_meta_data(filename=joiner_filename, meta_data=meta_data)
+    # trace the model
+    joiner_model.eval()
+    trace_model = torch.jit.trace(joiner_model, (projected_encoder_out, projected_decoder_out))
+    trace_model.save(joiner_filename)
+
+    mindietorch.set_device(0)
+    model = torch.jit.load(joiner_filename)
+    model.eval()
+
+    input_info = []
+    input_info.append(mindietorch.Input(projected_encoder_out.shape))
+    input_info.append(mindietorch.Input(projected_decoder_out.shape))
+    print('Start compiling model...')
+
+    compiled_model = mindietorch.compile(
+        model,
+        inputs=input_info,
+        precision_policy=_enums.PrecisionPolicy.FP16,
+        soc_version="Ascend310P3")
+    print('Joiner compiled successfully.')
+    save_path = str(joiner_filename).replace('.pt', f'_aie.pt')
+    compiled_model.save(save_path)
+
+    print(f'Compiled model saved to {save_path}')
+
+    # torch.onnx.export(
+    #     joiner_model,
+    #     (projected_encoder_out, projected_decoder_out),
+    #     joiner_filename,
+    #     verbose=False,
+    #     opset_version=opset_version,
+    #     input_names=[
+    #         "encoder_out",
+    #         "decoder_out",
+    #     ],
+    #     output_names=["logit"],
+    #     dynamic_axes={
+    #         "encoder_out": {0: "N"},
+    #         "decoder_out": {0: "N"},
+    #         "logit": {0: "N"},
+    #     },
+    # )
+    # meta_data = {
+    #     "joiner_dim": str(joiner_dim),
+    # }
+    # add_meta_data(filename=joiner_filename, meta_data=meta_data)


 @torch.no_grad()
@@ -719,7 +834,7 @@ def main():
     opset_version = 13

     logging.info("Exporting encoder")
-    encoder_filename = params.exp_dir / f"encoder-{suffix}.onnx"
+    encoder_filename = params.exp_dir / f"encoder-{suffix}.pt"
     export_encoder_model_onnx(
         encoder,
         encoder_filename,
@@ -728,7 +843,7 @@ def main():
     logging.info(f"Exported encoder to {encoder_filename}")

     logging.info("Exporting decoder")
-    decoder_filename = params.exp_dir / f"decoder-{suffix}.onnx"
+    decoder_filename = params.exp_dir / f"decoder-{suffix}.pt"
     export_decoder_model_onnx(
         decoder,
         decoder_filename,
@@ -737,7 +852,7 @@ def main():
     logging.info(f"Exported decoder to {decoder_filename}")

     logging.info("Exporting joiner")
-    joiner_filename = params.exp_dir / f"joiner-{suffix}.onnx"
+    joiner_filename = params.exp_dir / f"joiner-{suffix}.pt"
     export_joiner_model_onnx(
         joiner,
         joiner_filename,
@@ -748,31 +863,31 @@ def main():
     # Generate int8 quantization models
     # See https://onnxruntime.ai/docs/performance/model-optimizations/quantization.html#data-type-selection

-    logging.info("Generate int8 quantization models")
-
-    encoder_filename_int8 = params.exp_dir / f"encoder-{suffix}.int8.onnx"
-    quantize_dynamic(
-        model_input=encoder_filename,
-        model_output=encoder_filename_int8,
-        op_types_to_quantize=["MatMul"],
-        weight_type=QuantType.QInt8,
-    )
-
-    decoder_filename_int8 = params.exp_dir / f"decoder-{suffix}.int8.onnx"
-    quantize_dynamic(
-        model_input=decoder_filename,
-        model_output=decoder_filename_int8,
-        op_types_to_quantize=["MatMul", "Gather"],
-        weight_type=QuantType.QInt8,
-    )
-
-    joiner_filename_int8 = params.exp_dir / f"joiner-{suffix}.int8.onnx"
-    quantize_dynamic(
-        model_input=joiner_filename,
-        model_output=joiner_filename_int8,
-        op_types_to_quantize=["MatMul"],
-        weight_type=QuantType.QInt8,
-    )
+    # logging.info("Generate int8 quantization models")
+
+    # encoder_filename_int8 = params.exp_dir / f"encoder-{suffix}.int8.onnx"
+    # quantize_dynamic(
+    #     model_input=encoder_filename,
+    #     model_output=encoder_filename_int8,
+    #     op_types_to_quantize=["MatMul"],
+    #     weight_type=QuantType.QInt8,
+    # )
+
+    # decoder_filename_int8 = params.exp_dir / f"decoder-{suffix}.int8.onnx"
+    # quantize_dynamic(
+    #     model_input=decoder_filename,
+    #     model_output=decoder_filename_int8,
+    #     op_types_to_quantize=["MatMul", "Gather"],
+    #     weight_type=QuantType.QInt8,
+    # )
+
+    # joiner_filename_int8 = params.exp_dir / f"joiner-{suffix}.int8.onnx"
+    # quantize_dynamic(
+    #     model_input=joiner_filename,
+    #     model_output=joiner_filename_int8,
+    #     op_types_to_quantize=["MatMul"],
+    #     weight_type=QuantType.QInt8,
+    # )


 if __name__ == "__main__":
