--- attention_processor.py	2024-09-06 17:21:36.240822700 +0800
+++ attention_processor_patch.py	2024-09-06 17:50:07.631414800 +0800
@@ -2344,9 +2344,21 @@
 
         # the output of sdp = (batch, num_heads, seq_len, head_dim)
         # TODO: add support for attn.scale when we move to Torch 2.1
-        hidden_states = F.scaled_dot_product_attention(
-            query, key, value, attn_mask=attention_mask, dropout_p=0.0, is_causal=False
-        )
+        if query.device is not torch.device("cpu"):
+            if attention_mask is not None:
+                attention_mask=~attention_mask
+                attention_mask=attention_mask.to(torch.bool)
+            hidden_states=torch_npu.npu_prompt_flash_attention(query,key,value,
+                                                               atten_mask=attention_mask,
+                                                               input_layout='BNSD',
+                                                               scale_vaule=head_dim**-0.5,
+                                                               pre_tokens=65535,
+                                                               next_tokens=65535,
+                                                               num_heads=attn.heads)
+        else:    
+            hidden_states = F.scaled_dot_product_attention(
+                query, key, value, attn_mask=attention_mask, dropout_p=0.0, is_causal=False
+            )
 
         hidden_states = hidden_states.transpose(1, 2).reshape(batch_size, -1, attn.heads * head_dim)
         hidden_states = hidden_states.to(query.dtype)
