--- scheduling_cosine_dpmsolver_multistep.py	2024-09-12 09:03:22.789896000 +0800
+++ scheduling_cosine_dpmsolver_multistep_patch.py	2024-09-12 09:04:46.648920700 +0800
@@ -512,7 +512,7 @@
                     [g.initial_seed() for g in generator] if isinstance(generator, list) else generator.initial_seed()
                 )
             self.noise_sampler = BrownianTreeNoiseSampler(
-                model_output, sigma_min=self.config.sigma_min, sigma_max=self.config.sigma_max, seed=seed
+                model_output, sigma_min=self.config.sigma_min, sigma_max=self.config.sigma_max, seed=0
             )
         noise = self.noise_sampler(self.sigmas[self.step_index], self.sigmas[self.step_index + 1]).to(
             model_output.device
