--- transformer.py	2024-10-09 15:09:11.860795900 +0800
+++ transformer_patch.py	2024-10-09 15:15:51.763587200 +0800
@@ -500,34 +500,74 @@
             out = self.flash_attn(q, k, v, causal = causal, mask = final_attn_mask)
 
         else:
-            # Fall back to custom implementation
 
-            if h != kv_h:
+            batch, heads, q_len, _, k_len, device = *q.shape, k.shape[-2], q.device
+            kv_heads = k.shape[1]
+            # Recommended for multi-query single-key-value attention by Tri Dao
+            # kv shape torch.Size([1, 512, 64]) -> torch.Size([1, 8, 512, 64])
+
+            if heads != kv_heads:
                 # Repeat interleave kv_heads to match q_heads
-                heads_per_kv_head = h // kv_h
+                heads_per_kv_head = heads // kv_heads
                 k, v = map(lambda t: t.repeat_interleave(heads_per_kv_head, dim = 1), (k, v))
 
-            scale = 1. / (q.shape[-1] ** 0.5)
+            if k.ndim == 3:
+                k = rearrange(k, 'b ... -> b 1 ...').expand_as(q)
+
+            if v.ndim == 3:
+                v = rearrange(v, 'b ... -> b 1 ...').expand_as(q)
 
-            kv_einsum_eq = 'b j d' if k.ndim == 3 else 'b h j d'
+            causal = self.causal if causal is None else causal
 
-            dots = einsum(f'b h i d, {kv_einsum_eq} -> b h i j', q, k) * scale
+            if q_len == 1 and causal:
+                causal = False
             
-            i, j, dtype = *dots.shape[-2:], dots.dtype
+            if mask is not None:
+                assert mask.ndim == 4
+                mask = mask.expand(batch, heads, q_len, k_len)
+
+            # handle kv cache - this should be bypassable in updated flash attention 2
+
+            if k_len > q_len and causal:
+                causal_mask = self.create_causal_mask(q_len, k_len, device = device)
+                if mask is None:
+                    mask = ~causal_mask
+                else:
+                    mask = mask & ~causal_mask
+                causal = False
 
-            mask_value = -torch.finfo(dots.dtype).max
+            # manually handle causal mask, if another mask was given
 
-            if final_attn_mask is not None:
-                dots = dots.masked_fill(~final_attn_mask, mask_value)
+            row_is_entirely_masked = None
 
-            if causal:
-                causal_mask = self.create_causal_mask(i, j, device = device)
-                dots = dots.masked_fill(causal_mask, mask_value)
+            if mask is not None and causal:
+                causal_mask = self.create_causal_mask(q_len, k_len, device = device)
+                mask = mask & ~causal_mask
+
+                # protect against an entire row being masked out
+
+                row_is_entirely_masked = ~mask.any(dim = -1)
+                mask[..., 0] = mask[..., 0] | row_is_entirely_masked
+
+                causal = False
+            
+            if mask is not None:
+                mask =~ mask
+                mask = mask.to(torch.bool)
+            import torch_npu
+            out = torch_npu.npu_prompt_flash_attention(q, k, v,
+                                                       atten_mask=mask,
+                                                       input_layout='BNSD',
+                                                       scale_value=q.shape[-1]**-0.5,
+                                                       pre_tokens=65535,
+                                                       next_tokens=65535,
+                                                       num_heads=h,
+            )
 
-            attn = F.softmax(dots, dim=-1, dtype=torch.float32)
-            attn = attn.type(dtype)
+            # for a row that is entirely masked out, should zero out the output of that row token
 
-            out = einsum(f'b h i j, {kv_einsum_eq} -> b h i d', attn, v)
+            if row_is_entirely_masked is not None:
+                out = out.masked_fill(row_is_entirely_masked[..., None], 0.)
 
         # merge heads
         out = rearrange(out, ' b h n d -> b n (h d)')
