--- ./b/SenseVoice/model.py
+++ ./a/SenseVoice/model.py
@@ -271,10 +271,10 @@
 
     def forward(self, input):
         output = F.layer_norm(
-            input.float(),
+            input,
             self.normalized_shape,
-            self.weight.float() if self.weight is not None else None,
-            self.bias.float() if self.bias is not None else None,
+            self.weight if self.weight is not None else None,
+            self.bias if self.bias is not None else None,
             self.eps,
         )
         return output.type_as(input)
@@ -546,10 +546,9 @@
     def forward(
         self,
         xs_pad: torch.Tensor,
-        ilens: torch.Tensor,
+        masks: torch.Tensor,
     ):
         """Embed positions in tensor."""
-        masks = sequence_mask(ilens, device=ilens.device)[:, None, :]
 
         xs_pad *= self.output_size() ** 0.5
 
@@ -849,7 +848,12 @@
         speech_lengths += 3
 
         # Encoder
-        encoder_out, encoder_out_lens = self.encoder(speech, speech_lengths)
+        masks = sequence_mask(speech_lengths, device=speech_lengths.device)[:, None, :]
+        s = time.time()
+        encoder_out, encoder_out_lens = self.encoder(speech.to(torch.float16), masks.to(torch.float16))
+        e = time.time()
+        cost_time = e - s
+        meta_data["cost_time"] = cost_time
         if isinstance(encoder_out, tuple):
             encoder_out = encoder_out[0]
 
