--- attention_processor.py	2024-09-26 09:41:41.531223700 +0800
+++ attention_processor_patch.py	2024-09-26 09:52:34.363423900 +0800
@@ -2344,9 +2344,15 @@
 
         # the output of sdp = (batch, num_heads, seq_len, head_dim)
         # TODO: add support for attn.scale when we move to Torch 2.1
-        hidden_states = F.scaled_dot_product_attention(
-            query, key, value, attn_mask=attention_mask, dropout_p=0.0, is_causal=False
-        )
+        import mindietorch
+        scale_factor = query.size(-1)**-0.5
+        hidden_states = torch.ops.aie.flash_attention(query,key,value,
+                                                      num_head=attn.heads,
+                                                      attn_mask=attention_mask,
+                                                      pse=None,
+                                                      scale=scale_factor,
+                                                      layout="BNSD",
+                                                      type="PFA")
 
         hidden_states = hidden_states.transpose(1, 2).reshape(batch_size, -1, attn.heads * head_dim)
         hidden_states = hidden_states.to(query.dtype)
