--- brownian_interval.py	2024-09-06 18:31:27.596848500 +0800
+++ brownian_interval_patch.py	2024-09-06 18:39:44.069857000 +0800
@@ -226,7 +226,7 @@
             else:
                 # Don't compute space-time Levy area unless we need to
 
-                mean = left_diff * W * h_reciprocal
+                mean = left_diff * h_reciprocal * W
                 var = left_diff * right_diff * h_reciprocal
                 noise = parent._randn(parent._W_seed)
                 left_W = mean + math.sqrt(var) * noise
