+++ decoding.py
@@ -652,7 +652,10 @@
             # encoded audio features are given; skip audio encoding
             audio_features = mel
         else:
+            import time
+            time1 = time.time()
             audio_features = self.model.encoder(mel)
+            print(f"encode time = {(time.time() - time1) * 1000:.2f} ms")

         if audio_features.dtype != (
             torch.float16 if self.options.fp16 else torch.float32
@@ -683,6 +686,8 @@
         no_speech_probs = [np.nan] * n_batch

         try:
+            import time
+            time1 = time.time()
             for i in range(self.sample_len):
                 logits = self.inference.logits(tokens, audio_features)

@@ -703,6 +708,8 @@
                 tokens, completed = self.decoder.update(tokens, logits, sum_logprobs)

                 if completed or tokens.shape[-1] > self.n_ctx:
+                    avg_time = (time.time() - time1) / i * 1000
+                    print(f"avg decode time = {avg_time:.2f} ms")
                     break
         finally:
             self.inference.cleanup_caching()
@@ -824,3 +831,4 @@
     result = DecodingTask(model, options).run(mel)

     return result[0] if single else result
