--- attention_processor.py	2024-09-26 09:41:41.531223700 +0800
+++ attention_processor_patch.py	2024-10-16 15:56:30.103608300 +0800
@@ -2312,8 +2312,12 @@
         if kv_heads != attn.heads:
             # if GQA or MQA, repeat the key/value heads to reach the number of query heads.
             heads_per_kv_head = attn.heads // kv_heads
-            key = torch.repeat_interleave(key, heads_per_kv_head, dim=1)
-            value = torch.repeat_interleave(value, heads_per_kv_head, dim=1)
+            key = key.unsqueeze(2)
+            value = value.unsqueeze(2)
+            key = key.repeat(1, 1, heads_per_kv_head, 1, 1)
+            value = value.repeat(1, 1, heads_per_kv_head, 1, 1)
+            key = key.view(batch_size, attn.heads, sequence_length, head_dim)
+            value = value.view(batch_size, attn.heads, sequence_length, head_dim)
 
         if attn.norm_q is not None:
             query = attn.norm_q(query)
@@ -2344,9 +2348,15 @@
 
         # the output of sdp = (batch, num_heads, seq_len, head_dim)
         # TODO: add support for attn.scale when we move to Torch 2.1
-        hidden_states = F.scaled_dot_product_attention(
-            query, key, value, attn_mask=attention_mask, dropout_p=0.0, is_causal=False
-        )
+        import mindietorch
+        scale_factor = query.size(-1)**-0.5
+        hidden_states = torch.ops.aie.flash_attention(query,key,value,
+                                                      num_head=attn.heads,
+                                                      attn_mask=attention_mask,
+                                                      pse=None,
+                                                      scale=scale_factor,
+                                                      layout="BNSD",
+                                                      type="PFA")
 
         hidden_states = hidden_states.transpose(1, 2).reshape(batch_size, -1, attn.heads * head_dim)
         hidden_states = hidden_states.to(query.dtype)
