--- ./unet_2d_blocks.py	2024-06-24 09:06:04.593004325 +0800
+++ ./unet_2d_blocks_wz.py	2024-06-24 09:33:26.052027073 +0800
@@ -1159,6 +1159,7 @@ class CrossAttnDownBlock2D(nn.Module):
         attention_mask: Optional[torch.FloatTensor] = None,
         cross_attention_kwargs: Optional[Dict[str, Any]] = None,
         encoder_attention_mask: Optional[torch.FloatTensor] = None,
+        block_number: Optional[int]=None,
         additional_residuals: Optional[torch.FloatTensor] = None,
     ) -> Tuple[torch.FloatTensor, Tuple[torch.FloatTensor, ...]]:
         output_states = ()
@@ -1210,6 +1211,8 @@ class CrossAttnDownBlock2D(nn.Module):
                 hidden_states = hidden_states + additional_residuals

             output_states = output_states + (hidden_states,)
+            if block_number is not None and len(output_states) == block_number + 1:
+                return hidden_states, output_states

         if self.downsamplers is not None:
             for downsampler in self.downsamplers:
@@ -2364,6 +2367,7 @@ class CrossAttnUpBlock2D(nn.Module):
         upsample_size: Optional[int] = None,
         attention_mask: Optional[torch.FloatTensor] = None,
         encoder_attention_mask: Optional[torch.FloatTensor] = None,
+        block_number: Optional[int]=None,
     ) -> torch.FloatTensor:
         lora_scale = cross_attention_kwargs.get("scale", 1.0) if cross_attention_kwargs is not None else 1.0
         is_freeu_enabled = (
@@ -2372,8 +2376,12 @@ class CrossAttnUpBlock2D(nn.Module):
             and getattr(self, "b1", None)
             and getattr(self, "b2", None)
         )
-
-        for resnet, attn in zip(self.resnets, self.attentions):
+
+        prev_feature = []
+        for i, (resnet, attn) in enumerate(zip(self.resnets, self.attentions)):
+            if block_number is not None and i < len(self.resnets) - block_number - 1:
+                continue
+
             # pop res hidden states
             res_hidden_states = res_hidden_states_tuple[-1]
             res_hidden_states_tuple = res_hidden_states_tuple[:-1]
@@ -2390,8 +2398,9 @@ class CrossAttnUpBlock2D(nn.Module):
                     b2=self.b2,
                 )

+            prev_feature.append(hidden_states)
             hidden_states = torch.cat([hidden_states, res_hidden_states], dim=1)
-
+
             if self.training and self.gradient_checkpointing:

                 def create_custom_forward(module, return_dict=None):
@@ -2428,12 +2437,12 @@ class CrossAttnUpBlock2D(nn.Module):
                     encoder_attention_mask=encoder_attention_mask,
                     return_dict=False,
                 )[0]
-
+
         if self.upsamplers is not None:
             for upsampler in self.upsamplers:
                 hidden_states = upsampler(hidden_states, upsample_size, scale=lora_scale)
-
-        return hidden_states
+
+        return hidden_states, prev_feature


 class UpBlock2D(nn.Module):
