diff --git a/egs/librispeech/ASR/zipformer/export-onnx-streaming.py b/egs/librispeech/ASR/zipformer/export-onnx-streaming.py
index 6bc9b18..345e0c9 100755
--- a/egs/librispeech/ASR/zipformer/export-onnx-streaming.py
+++ b/egs/librispeech/ASR/zipformer/export-onnx-streaming.py
@@ -273,7 +273,7 @@ class OnnxEncoder(nn.Module):
 
         states.append(embed_states)
 
-        processed_lens = torch.zeros(batch_size, dtype=torch.int64, device=device)
+        processed_lens = torch.zeros(batch_size, dtype=torch.int32, device=device)
         states.append(processed_lens)
 
         return states
@@ -505,7 +505,7 @@ def export_decoder_model_onnx(
     context_size = decoder_model.decoder.context_size
     vocab_size = decoder_model.decoder.vocab_size
 
-    y = torch.zeros(10, context_size, dtype=torch.int64)
+    y = torch.zeros(10, context_size, dtype=torch.int32)
     decoder_model = torch.jit.script(decoder_model)
     torch.onnx.export(
         decoder_model,
@@ -748,31 +748,31 @@ def main():
     # Generate int8 quantization models
     # See https://onnxruntime.ai/docs/performance/model-optimizations/quantization.html#data-type-selection
 
-    logging.info("Generate int8 quantization models")
-
-    encoder_filename_int8 = params.exp_dir / f"encoder-{suffix}.int8.onnx"
-    quantize_dynamic(
-        model_input=encoder_filename,
-        model_output=encoder_filename_int8,
-        op_types_to_quantize=["MatMul"],
-        weight_type=QuantType.QInt8,
-    )
-
-    decoder_filename_int8 = params.exp_dir / f"decoder-{suffix}.int8.onnx"
-    quantize_dynamic(
-        model_input=decoder_filename,
-        model_output=decoder_filename_int8,
-        op_types_to_quantize=["MatMul", "Gather"],
-        weight_type=QuantType.QInt8,
-    )
-
-    joiner_filename_int8 = params.exp_dir / f"joiner-{suffix}.int8.onnx"
-    quantize_dynamic(
-        model_input=joiner_filename,
-        model_output=joiner_filename_int8,
-        op_types_to_quantize=["MatMul"],
-        weight_type=QuantType.QInt8,
-    )
+    # logging.info("Generate int8 quantization models")
+    #
+    # encoder_filename_int8 = params.exp_dir / f"encoder-{suffix}.int8.onnx"
+    # quantize_dynamic(
+    #     model_input=encoder_filename,
+    #     model_output=encoder_filename_int8,
+    #     op_types_to_quantize=["MatMul"],
+    #     weight_type=QuantType.QInt8,
+    # )
+    #
+    # decoder_filename_int8 = params.exp_dir / f"decoder-{suffix}.int8.onnx"
+    # quantize_dynamic(
+    #     model_input=decoder_filename,
+    #     model_output=decoder_filename_int8,
+    #     op_types_to_quantize=["MatMul", "Gather"],
+    #     weight_type=QuantType.QInt8,
+    # )
+    #
+    # joiner_filename_int8 = params.exp_dir / f"joiner-{suffix}.int8.onnx"
+    # quantize_dynamic(
+    #     model_input=joiner_filename,
+    #     model_output=joiner_filename_int8,
+    #     op_types_to_quantize=["MatMul"],
+    #     weight_type=QuantType.QInt8,
+    # )
 
 
 if __name__ == "__main__":
