---
 models/blip.py | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/models/blip.py b/models/blip.py
index 38678f6..a1c1771 100644
--- a/models/blip.py
+++ b/models/blip.py
@@ -126,7 +126,10 @@ class BLIP_Decoder(nn.Module):
         return loss_lm
         
     def generate(self, image, sample=False, num_beams=3, max_length=30, min_length=10, top_p=0.9, repetition_penalty=1.0):
-        image_embeds = self.visual_encoder(image)
+        if 'om' in dir(self.visual_encoder):
+            image_embeds = torch.from_numpy(self.visual_encoder.om.infer([image.numpy()])[0])
+        else:
+            image_embeds = self.visual_encoder(image)
 
         if not sample:
             image_embeds = image_embeds.repeat_interleave(num_beams,dim=0)
-- 
