diff --git a/src/transformers/modeling_utils.py b/src/transformers/modeling_utils.py
index 152e52fe2..67b7a6bb7 100644
--- a/src/transformers/modeling_utils.py
+++ b/src/transformers/modeling_utils.py
@@ -788,7 +788,7 @@ class ModuleUtilsMixin:
         # Since we are adding it to the raw scores before the softmax, this is
         # effectively the same as removing these entirely.
         extended_attention_mask = extended_attention_mask.to(dtype=dtype)  # fp16 compatibility
-        extended_attention_mask = (1.0 - extended_attention_mask) * torch.finfo(dtype).min
+        extended_attention_mask = (1.0 - extended_attention_mask) * (-1000.0)
         return extended_attention_mask
 
     def get_head_mask(
diff --git a/src/transformers/trainer.py b/src/transformers/trainer.py
index ac166f002..7c869f002 100755
--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -53,6 +53,7 @@ from .integrations import (  # isort: split
 
 import numpy as np
 import torch
+import torch_aie
 import torch.distributed as dist
 from packaging import version
 from torch import nn
@@ -204,7 +205,8 @@ TRAINER_STATE_NAME = "trainer_state.json"
 OPTIMIZER_NAME = "optimizer.pt"
 SCHEDULER_NAME = "scheduler.pt"
 SCALER_NAME = "scaler.pt"
-
+NPU_DEVICE = 'npu:0'
+torch_aie.set_device(int(NPU_DEVICE.split(':')[-1]))
 
 class Trainer:
     """
@@ -327,6 +329,10 @@ class Trainer:
         # force device and distributed setup init explicitly
         args._setup_devices
 
+        # calculate torch aie inference time
+        self.count = 0
+        self.inference_times = []
+
         if model is None:
             if model_init is not None:
                 self.model_init = model_init
@@ -1257,17 +1263,24 @@ class Trainer:
                 example_tensor = torch.ones_like(example_batch[key])
                 jit_inputs.append(example_tensor)
             jit_inputs = tuple(jit_inputs)
+            is_jit_model = False
             try:
                 jit_model = model.eval()
                 with ContextManagers([self.autocast_smart_context_manager(), torch.no_grad()]):
-                    jit_model = torch.jit.trace(jit_model, jit_inputs, strict=False)
+                    jit_model = torch.jit.trace(jit_model, jit_inputs[:3], strict=False)
                 jit_model = torch.jit.freeze(jit_model)
-                jit_model(**example_batch)
+                example_batch.pop("labels")
+                input_ids = example_batch["input_ids"]
+                attention_mask = example_batch["attention_mask"]
+                token_type_ids = example_batch["token_type_ids"]
+                jit_model(input_ids, attention_mask, token_type_ids)
                 model = jit_model
+                is_jit_model = True
+                logger.info("torch jit model success.")
             except (RuntimeError, TypeError) as e:
                 logger.warning(f"failed to use PyTorch jit mode due to: {e}.")
 
-        return model
+        return model, is_jit_model
 
     def ipex_optimize_model(self, model, training=False, dtype=torch.float32):
         if not is_ipex_available():
@@ -1291,13 +1304,37 @@ class Trainer:
 
         return model
 
+    def torch_aie_compile(self, dataloader, ts_model):
+        max_seq_length = 512
+        input_info = [
+            torch_aie.Input((dataloader.batch_size, max_seq_length), dtype=torch.int64),
+            torch_aie.Input((dataloader.batch_size, max_seq_length), dtype=torch.int64),
+            torch_aie.Input((dataloader.batch_size, max_seq_length), dtype=torch.int64)
+            ]
+        model = torch_aie.compile(
+            ts_model,
+            inputs=input_info,
+            precision_policy=torch_aie.PrecisionPolicy.FP16,
+            truncate_long_and_double=True,
+            require_full_compilation=False,
+            allow_tensor_replace_int=True,
+            torch_executed_ops=[],
+            soc_version="Ascend310P3",
+            optimization_level=0
+        )
+        model.eval()
+        return model
+
     def _wrap_model(self, model, training=True, dataloader=None):
         if self.args.use_ipex:
             dtype = torch.bfloat16 if self.use_cpu_amp else torch.float32
             model = self.ipex_optimize_model(model, training, dtype=dtype)
 
         if self.args.jit_mode_eval:
-            model = self.torch_jit_model_eval(model, dataloader, training)
+            model, is_jit_model = self.torch_jit_model_eval(model, dataloader, training)
+            if is_jit_model and self.args.torch_aie_enable:
+                torch_aie.set_device(0)
+                model = self.torch_aie_compile(dataloader, model)
 
         if is_sagemaker_mp_enabled():
             # Wrapping the base model twice in a DistributedModel will raise an error.
@@ -2994,6 +3031,7 @@ class Trainer:
                     else nested_concat(inputs_host, inputs_decode, padding_index=-100)
                 )
             if logits is not None:
+                logits = logits.view(batch_size, 512, 9)
                 logits = self._pad_across_processes(logits)
                 logits = self._nested_gather(logits)
                 if self.preprocess_logits_for_metrics is not None:
@@ -3024,6 +3062,11 @@ class Trainer:
 
                 # Set back to None to begin a new accumulation
                 losses_host, preds_host, inputs_host, labels_host = None, None, None, None
+        if args.torch_aie_enable:
+            average_inference_time = np.mean(self.inference_times)
+            logger.info(f"Pure model inference performance per sample = {average_inference_time * 1000} ms")
+            logger.info(f"Performance= {batch_size * 1.0 / average_inference_time} it/s")
+            logger.info(f"inference time count/total count: {len(self.inference_times)} / {self.count}")
 
         if args.past_index and hasattr(self, "_past"):
             # Clean the state at the end of the evaluation loop
@@ -3211,6 +3254,24 @@ class Trainer:
                     else:
                         logits_mb = raw_outputs
                     logits = smp_nested_concat(logits_mb)
+            elif self.args.torch_aie_enable and has_labels:
+                loss = None
+                logger.info("torch aie model forward... ")
+                inputs_npu1 = inputs["input_ids"].to(NPU_DEVICE)
+                inputs_npu2 = inputs["attention_mask"].to(NPU_DEVICE)
+                inputs_npu3 = inputs["token_type_ids"].to(NPU_DEVICE)
+                stream = torch_aie.npu.Stream(NPU_DEVICE)
+                with torch_aie.npu.stream(stream):
+                    inf_start = time.time()
+                    outputs = model(inputs_npu1, inputs_npu2, inputs_npu3)
+                    stream.synchronize()
+                    inf_end = time.time()
+                    inf_time = inf_end - inf_start
+                    self.count += 1
+                    if self.count > 5:
+                        self.inference_times.append(inf_time)
+                logits = outputs["logits"]
+                logger.info(f"step: {self.count} infer time: {inf_time}s")
             else:
                 if has_labels:
                     with self.compute_loss_context_manager():
@@ -3506,11 +3567,11 @@ class Trainer:
 
         # if full fp16 or bf16 eval is wanted and this ``evaluation`` or ``predict`` isn't called
         # while ``train`` is running, cast it to the right dtype first and then put on device
-        if not self.is_in_train:
-            if args.fp16_full_eval:
-                model = model.to(dtype=torch.float16, device=args.device)
-            elif args.bf16_full_eval:
-                model = model.to(dtype=torch.bfloat16, device=args.device)
+        # if not self.is_in_train:
+        #     if args.fp16_full_eval:
+        #         model = model.to(dtype=torch.float16, device=args.device)
+        #     elif args.bf16_full_eval:
+        #         model = model.to(dtype=torch.bfloat16, device=args.device)
 
         batch_size = dataloader.batch_size
         num_examples = self.num_examples(dataloader)
diff --git a/src/transformers/training_args.py b/src/transformers/training_args.py
index fc5ace752..19df2cfce 100644
--- a/src/transformers/training_args.py
+++ b/src/transformers/training_args.py
@@ -995,6 +995,7 @@ class TrainingArguments:
             "help": "Overrides the default timeout for distributed training (value should be given in seconds)."
         },
     )
+    torch_aie_enable: bool = field(default=True, metadata={"help": "Whether or not to use torch aie."})
 
     def __post_init__(self):
         # Handle --use_env option in torch.distributed.launch (local_rank not passed as an arg then).
