---
 models/med.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/models/med.py b/models/med.py
index 7b00a35..eb3131f 100644
--- a/models/med.py
+++ b/models/med.py
@@ -904,14 +904,14 @@ class BertLMHeadModel(BertPreTrainedModel):
         prediction_scores = self.cls(sequence_output)
         
         if return_logits:
-            return prediction_scores[:, :-1, :].contiguous()  
+            return prediction_scores[:, 0, :]
 
         lm_loss = None
         if labels is not None:
             # we are doing next-token prediction; shift prediction scores and input ids by one
             shifted_prediction_scores = prediction_scores[:, :-1, :].contiguous()
             labels = labels[:, 1:].contiguous()
-            loss_fct = CrossEntropyLoss(reduction=reduction, label_smoothing=0.1) 
+            loss_fct = CrossEntropyLoss(reduction=reduction)
             lm_loss = loss_fct(shifted_prediction_scores.view(-1, self.config.vocab_size), labels.view(-1))
             if reduction=='none':
                 lm_loss = lm_loss.view(prediction_scores.size(0),-1).sum(1)               
-- 
